// angular-select/lib/multi-select.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var MultiSelectComponent = MultiSelectComponent_1 = (function () {
    function MultiSelectComponent() {
        this.config = { allowNewOptions: false, closeOnSelect: false, deleteOnBackspace: true };
        this.emptyText = '';
        this.emptyOptions = '';
        this.change = new core_1.EventEmitter();
        this.onChangeByUser = new core_1.EventEmitter();
        this._activeOptionIndex = 0;
        this._selections = [];
        this._options = [];
        this.keyup$ = new BehaviorSubject_1.BehaviorSubject(null);
        this.propagateChange = function (_) { };
        this.propagateTouched = function () { };
    }
    Object.defineProperty(MultiSelectComponent.prototype, "selections", {
        get: function () {
            return this._selections;
        },
        set: function (selections) {
            this._selections = isStringArray(selections) ? selections : [];
            this.propagateChange(this.selections);
            this.change.emit(selections);
        },
        enumerable: true,
        configurable: true
    });
    MultiSelectComponent.prototype.delete = function ($event, selection) {
        if (this.disabled)
            return;
        $event.preventDefault();
        var idx = this.selections.indexOf(selection);
        this.selections = this.selections
            .slice(0, idx)
            .concat(this.selections.slice(idx + 1, this.selections.length));
        this.onChangeByUser.emit(this.selections);
    };
    MultiSelectComponent.prototype.focus = function ($event) {
        if (this.disabled)
            return;
        if ($event.type === 'focusin') {
            this.isFocused = true;
            this.showOptions = true;
        }
        else {
            this.isFocused = false;
            this.showOptions = false;
            this.propagateTouched();
        }
    };
    MultiSelectComponent.prototype.getEmptyText = function () {
        return this.selections.length === 0
            ? this.emptyText || ''
            : '';
    };
    MultiSelectComponent.prototype.keydown = function ($event) {
        if ($event.key === 'Enter' && this.showOptions) {
            $event.preventDefault();
        }
        if ($event.key === 'Backspace' && this.config.deleteOnBackspace && this.inputRef.nativeElement.value === '') {
            this.selections = this.selections.slice(0, -1);
            this.onChangeByUser.emit(this.selections);
        }
    };
    MultiSelectComponent.prototype.keyup = function ($event) {
        this.keyup$.next($event);
        switch ($event.key) {
            case 'Escape':
                this.showOptions = false;
                break;
            default:
                this.inputValue = this.inputRef.nativeElement.value;
                break;
        }
    };
    MultiSelectComponent.prototype.registerOnChange = function (fn) {
        this.propagateChange = fn;
    };
    MultiSelectComponent.prototype.registerOnTouched = function (fn) {
        this.propagateTouched = fn;
    };
    MultiSelectComponent.prototype.select = function (option) {
        if (!option)
            return;
        if (this.config.closeOnSelect)
            this.showOptions = false;
        this.inputRef.nativeElement.value = '';
        this.inputValue = '';
        this.selections = this.selections.concat(option);
        this.onChangeByUser.emit(this.selections);
    };
    MultiSelectComponent.prototype.setDisabledState = function (disabled) {
        this.disabled = disabled;
    };
    MultiSelectComponent.prototype.writeValue = function (value) {
        if (value === void 0) { value = []; }
        this.selections = value;
    };
    return MultiSelectComponent;
}());
__decorate([
    core_1.Input(),
    __metadata("design:type", Object)
], MultiSelectComponent.prototype, "config", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", String)
], MultiSelectComponent.prototype, "emptyText", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", String)
], MultiSelectComponent.prototype, "emptyOptions", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Array)
], MultiSelectComponent.prototype, "options", void 0);
__decorate([
    core_1.Output(),
    __metadata("design:type", Object)
], MultiSelectComponent.prototype, "change", void 0);
__decorate([
    core_1.Output(),
    __metadata("design:type", Object)
], MultiSelectComponent.prototype, "onChangeByUser", void 0);
__decorate([
    core_1.ViewChild('inputRef'),
    __metadata("design:type", core_1.ElementRef)
], MultiSelectComponent.prototype, "inputRef", void 0);
MultiSelectComponent = MultiSelectComponent_1 = __decorate([
    core_1.Component({
        selector: 'ng-multi-select',
        template: '<div class="select-input" [class.active]="isFocused" [attr.disabled]="disabled ? true : null"><div *ngFor="let selection of selections"><a (mousedown)="delete($event, selection)"><span>{{ selection }}</span> x</a></div><input type="text" (click)="focus({ type: \'focusin\' })" (keyup)="keyup($event)" (keydown)="keydown($event)" (focusin)="focus($event)" (focusout)="focus($event)" [placeholder]="getEmptyText()" [attr.disabled]="disabled ? true : null" #inputRef></div><options [keyup]="keyup$" [emptyOptions]="emptyOptions" [options]="options" [selections]="selections" [showOptions]="showOptions" [inputValue]="inputValue" (onSelect)="select($event)"></options>',
        styles: [':host{display:inline-block;position:relative}.select-input{display:flex;flex-wrap:wrap}.select-input>input{display:block;min-width:50%;flex-grow:1}'],
        providers: [{
                provide: forms_1.NG_VALUE_ACCESSOR,
                useExisting: core_1.forwardRef(function () { return MultiSelectComponent_1; }),
                multi: true,
            }]
    })
], MultiSelectComponent);
exports.MultiSelectComponent = MultiSelectComponent;
function isStringArray(obj) {
    if (obj === null)
        return false;
    if (!Array.isArray(obj))
        return false;
    return obj.every(function (value) { return typeof value === 'string'; });
}
var MultiSelectComponent_1;
